### PR TITLE
Implicitly restore C++/PackageReference in Visual Studio

### PR DESCRIPTION
**Pull request by JunielKatarn**
There is still refusal to officially support building `PackageReference` for C++ projects in Visual Studio even though the builds can be produced using `MSBuild.exe` since around version 16.7 (command line).

See https://developercommunity.visualstudio.com/t/use-packagereference-in-vcxproj/351636?from=email&viewtype=all#T-ND1617002

This change enforces the restore target for all C++ projects if the build is initiated from the IDE (i.e. `Ctrl+Shift+B`), emulating the following command:
```
MSBuild.exe /restore <project or solution>
```

Note, .NET/C# projects do need/use this change because Visual Studio does support implicit `PackageReference` restore during build.

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/microsoft/react-native-windows/pull/9303)

### CODE REVIEW

File: Directory.Build.props [hunk 1]
```diff
@@ -43,10 +43,13 @@
     <TargetFrameworkMoniker>native,Version=v0.0</TargetFrameworkMoniker>
     <!-- TODO: Keep? -->
     <!-- <RuntimeIdentifiers Condition="'$(RuntimeIdentifiers)' == ''">win;win-x64;win-x86;win-arm64</RuntimeIdentifiers> -->
+
+    <!--See https://docs.microsoft.com/en-us/nuget/reference/msbuild-targets#restore-target-->
```
Comment (NickGerleman) [inline Directory.Build.props:47]: These props only apply to our repo. Can we put them somewhere users will also get them?

Comment (JunielKatarn) [reply Directory.Build.props:47]: Would `vnext\Directory.Build.*` work, or should this happen somewhere in `vnext\PropertySheets\External\...`?

Comment (NickGerleman) [reply Directory.Build.props]: External property sheets are used for projects external to Microsoft.ReactNative.dll. They are included in the projects owned by the individual apps.

Imagine projects like `playground` or `e2e-test-app`.

**Comment by NickGerleman** [comment]
Just as a heads up, I'm going not going to be looking much at PRs for the next couple of weeks. I won't be looking at PRS too often, or too closely. It's probably worth some brainstorming with RNW NuGet folks to get everyone on the same page.

I'd appreciate though, if you could keep your ears open on anyone who may be running into build issues. The current error doesn't make it immediately obvious that the failure is due to missing NuGet packages.

**Comment by JunielKatarn** [comment]
Supersedes #9301.

**Comment by NickGerleman** [comment]
I've been thinking about this. Asking Visual Studio to use MSBuild for restoration is a good change. C# RNW apps have run into issues where `PackageReference` behavior is incorrect in Visual Studio, so taking the slightly slower more correct path seems preferable (potentially even in C# projects). We already have VS rely on MSBuild incremental build instead of VS incremental build, for similar reasons.

**Comment by NickGerleman** [comment]
In an end-to-end RNW application, the `vnext` props roughly corresponds to our DLL and everything internal to it. The PackageReference change applies to this, but also applies to user's projects in Visual Studio, that includes our DLL.

`External` property sheets apply to RNW projects that consume the MSRN DLL. We have some examples of them in separate packages in our repo. E.g. `playground`, `e2e-test-app`, `integration-test-app`, `sample-apps`.

The PackageReference change applies to these projects, so we should include any related changes in the external property sheets as well.

**Review by NickGerleman — APPROVED** [review]

File: Directory.Build.props (hunk 1 above)
Comment (JunielKatarn) [reply Directory.Build.props]: The templates were updated to use `PackageReference`.
I do notice that for C#, we copy the `vnext\Directory.Build.*` files into the template.
I'll do the same for the C++ project templates (those props/targets contain the VS restore logic) and merge if tests pass.

**Pull request merged**